src/convnets/convnext.jl

```diff
@@ -0,0 +1,104 @@
+"""
+    convnextblock(planes, drop_path = 0., λ = 1f-6)
+
+Creates a single block of ConvNeXt.
+([reference](https://arxiv.org/abs/2201.03545))
+
+# Arguments:
+- `planes`: number of input channels.
+- `drop_path_rate`: Stochastic depth rate.
+- `λ`: Init value for [LayerScale](https://arxiv.org/abs/2103.17239)
+"""
+function convnextblock(planes, drop_path_rate = 0., λ = 1f-6)
+  γ = λ > 0 ? Flux.ones32(planes) * λ : nothing
+  LayerScale(x) = isnothing(γ) ? x : x .* γ
```

Comment:
What is this doing? Is the `γ` array meant to be trainable? I don't think `params` (or `functor`) will see it. 

If not, can't this just be `x -> x .* λ`?

Reply:
Yep, it's intended to be learnable 😬 How do I make it visible to the `params` and `functor`?

Reply:
I think it needs to be a struct, which you mark with `@functor`.

What you want is almost Flux.Diagonal, except without the β. It would be very easy to make `Flux.Diagonal(fill!(ones32(planes), λ), false)` work, but it doesn't right now as the signature is `Flux.Diagonal(::T, ::T)`.

Reply:
> What you want is almost Flux.Diagonal, except without the β.

That's exactly what I want 😅 LayerScale is defined as 

"(it)...adds a learnable diagonal matrix on output of each residual block, initialised close to (but not at) 0. Adding this simple layer after each residual block improves the training dynamic, allowing for the training of deeper high-capacity image transformers that benefit from depth."

> it doesn't right now as the signature is `Flux.Diagonal(::T, ::T).`

Maybe a modification and an addition constructor to make the bias optional? I can open a PR if that's okay - this seems like something that would probably be useful in a lot of places

Reply:
Yes, maybe this is spelled `Diagonal(init=λ, bias=false)` with `init::Number` understood as a scale?

Some chance others may prefer this to be a separate `Scale` layer --- the bias here can be the full size, not just a vector, but whether that makes it more or less like Dense is perhaps a matter of interpretation.

Reply:
In a sense `Dense` already handles this, but is somewhat poorly named (c.f. the example `Affine` layer in the docs or `Linear`). It also misses out on the fused broadcast, but that may be a non-issue given Zygote unfuses it anyhow.

Reply:
Not sure we have to special case `Number`, but a PR to Flux to make `Diagonal` accept the same initialization/hardcoded parameters in the constructor as everything else sounds like the right step to take here.

Reply:
> Dense already handles this

What do you mean? This wants `w .* x` with some trainable array `w = fill(λ, ...)`, while Dense does `w * x`. 

> the same initialization/hardcoded parameters

But Diagonal doesn't initialise the bias the same way: It's not just a vector across channels, it's an array the same size as the other.

Reply:
Tbh, "diagonal" makes most sense to me. `Dense` is an affine transformation mathematically speaking. `Diagonal` is an affine transformation where the matrix is constrained to be diagonal.

> But Diagonal doesn't initialise the bias the same way: It's not just a vector across channels, it's an array the same size as the other.

Couldn't this be similar to `Dense` where `create_bias` is called with the correct dims? I agree it's not the same shape as the others, so we should not initialize it that way. I meant `Diagonal` should allow bias to be:
- the array itself
- `false`/`true`
Similarly, `Diagonal` should accept an `init` for the weight vector.

Reply:
Re Dense, sure you could allow just a number, but here if I understood right we only start with a constant scale like that; afterwards each may be learned independently.

Seems a poor choice that Flux.Diagonal clashes with LinearAlgebra. Maybe we should even change it when doing these upgrades.

Reply:
It's an interesting thought, maybe in the future `Diagonal` could literally be `Dense` with a `Diagonal` weight matrix. What I haven't wrapped my head around is how all of this would function with >2D inputs, but that's not important for this PR.

Reply:
If I've understood this correctly, then, I should open a PR for a modified version of Diagonal which makes the bias optional? Maybe the name can be discussed there further too

Reply:
No rush I think, we can get this PR through first.

Reply:
Cool, then I'm leaving this part untouched for now. Some feedback on `ChannelLayerNorm` would be appreciated too, that one's been eating my head up far more than this

Reply:
This part still needs to be updated so that `params` captures the trainable diagonal. It would need to be changed to something like https://github.com/FluxML/Metalhead.jl/pull/119#discussion_r809195465 except instead of `false` you can use `Zeros()`. This would be what th

Reply:
This part still needs to be updated so that `params` captures the trainable diagonal. It would need to be changed to something like https://github.com/FluxML/Metalhead.jl/pull/119#discussion_r809195465 except instead of `false` you can use `Zeros()`. Having a convenience constructor for this can be done later than this PR.

Reply:
I checked it out and the `show` function seems to be capturing the diagonal and it's parameters, and the number of total params right now check out with the original model, so somehow it's already working the way it's intended to